### DISCUSSION TITLE
Realtime not sending notifications

### DISCUSSION THREAD

**Discussion by jvbraganca** [supabase/supabase#428] [Questions] [answered]
Hello,

I am trying to receive notifications in the realtime but when making changes or inserting them into the DB I am not receiving them.

Here is how I declared Supabase:

./assets/javascript/supabase
```js
import { createClient } from '@supabase/supabase-js'

export const supabase = createClient(process.env.SUPABASE_API_URL, process.env.SUPABASE_API_KEY)
```
Here is how I am calling it from my component
```js
import { supabase } from '@/assets/javascript/supabase'
subscribeToNotifications () {
      supabase.from('notifications').on('*', (payload) => {
        // this.appendToMessages(payload.new)
        console.log(payload)
      }).subscribe()

      const subscriptions = supabase.getSubscriptions()
      console.log(subscriptions)
    },
```
I am connecting to the socket and receiving the connection when I output the `supabase.getSubscriptions()`
But, when I make changes to the table I cannot receive them from the realtime. 
The connection is fine as I can insert and read data through the normal API. Can someone give me some light, please?

**Comment by kiwicopple** [comment, marked as answer]
Hi @jvbraganca, it looks fine to me. Perhaps you try listening to your whole database and let me know if receive anything?

```js
const mySubscription = supabase
  .from('*')
  .on('*', payload => {
    console.log('Change received!', payload)
  })
  .subscribe()
```

Also, can you log the result of `mySubscription` and let us know what it is?

```js
console.log('mySubscription', mySubscription)
```

**Reply by jvbraganca** [reply]
Hello, @kiwicopple I tried to listen to all changes on the database but still haven't got any output. Here is the `console.log('mySubscription', mySubscription)
`:

```js
RealtimeSubscription
bindings: Array(6)
0: {event: "phx_close", callback: ƒ}
1: {event: "phx_error", callback: ƒ}
2: {event: "phx_reply", callback: ƒ}
3: {event: "*", callback: ƒ}
4: {event: "phx_error", callback: ƒ}
5: {event: "phx_close", callback: ƒ}
length: 6
__proto__: Array(0)
joinPush: Push
channel: RealtimeSubscription {topic: "realtime:public", params: {…}, socket: RealtimeClient, bindings: Array(6), state: "joined", …}
event: "phx_join"
payload: {}
recHooks: (5) [{…}, {…}, {…}, {…}, {…}]
receivedResp: {response: {…}, status: "ok"}
ref: "1"
refEvent: "chan_reply_1"
sent: true
timeout: 10000
timeoutTimer: undefined
__proto__: Object
joinedOnce: true
params: {}
pushBuffer: []
rejoinTimer: Timer {timer: undefined, tries: 0, callback: ƒ, timerCalc: ƒ}
socket: RealtimeClient
channels: [RealtimeSubscription]
conn: WebSocket {url: "wss://tbqtyripmdyrulzpwzvb.supabase.co/realtime/v1…v0xSJdK70kRvQsijNpDNjlw_f8XWB0jfOPyFCeM&vsn=1.0.0", readyState: 1, bufferedAmount: 0, onopen: ƒ, onerror: ƒ, …}
decode: (payload, callback) => {…}
encode: (payload, callback) => {…}
endPoint: "wss://PROJECT_ID.supabase.co/realtime/v1/websocket"
headers: {}
heartbeatIntervalMs: 30000
heartbeatTimer: 88
logger: () => { }
longpollerTimeout: 20000
params: {apikey: "API_KEY"}
pendingHeartbeatRef: null
reconnectAfterMs: (tries) => {…}
reconnectTimer: Timer {timer: undefined, tries: 0, callback: ƒ, timerCalc: ƒ}
ref: 3
sendBuffer: []
stateChangeCallbacks: {open: Array(0), close: Array(0), error: Array(0), message: Array(0)}
timeout: 10000
transport: ƒ W3CWebSocket(uri, protocols)
__proto__: Object
state: "joined"
timeout: 10000
topic: "realtime:public"
```

**Reply by kiwicopple** [reply]
Strange. Did you restore your database from a dump perhaps?

Can you show me the output of this? `select * from pg_replication_slots;`

If there is nothing from that request, can you try running `CREATE PUBLICATION supabase_realtime FOR ALL TABLES;`?

**Reply by jvbraganca** [reply]
No, it's a new database.

`select * from pg_replication_slots;` outputs:
![image](https://user-images.githubusercontent.com/16918085/103011316-789cb680-4518-11eb-845d-5ac400941f9c.png)

`CREATE PUBLICATION supabase_realtime FOR ALL TABLES;` outputs:
`ERROR: publication "supabase_realtime" already exists`

**Reply by kiwicopple** [reply]
I'm sorry then @jvbraganca - I'm a bit stumped. I'm guessing it has soemthing to do with the way your code is set up. Could you either:

- start a new supabase project and try with that (to eliminate the possibility that it's a database problem)
- create a https://codesandbox.io/ to replicate the issue?

**Reply by jvbraganca** [reply]
@kiwicopple  I created a new database for the CodeBox but I am still not able to receive the data from the subscription. Here is the output for the project I have created:

`select * from pg_replication_slots;` outputs:
![image](https://user-images.githubusercontent.com/16918085/103015561-6a05cd80-451f-11eb-9067-1c258d6a5e7e.png)

`CREATE PUBLICATION supabase_realtime FOR ALL TABLES;` outputs:

`ERROR: publication "supabase_realtime" already exists`

#### Here is the link for the CodeBox: https://codesandbox.io/s/supabase-subscription-error-joz8i?file=/src/App.vue

**Comment by jvbraganca** [comment]
Anything?

**Reply by soedirgo** [reply]
Hey @jvbraganca, I tried the codesandbox and it gave `426 Upgrade Required` errors, probably a limitation with codesandbox.

But I downloaded the code and run it locally and it seems to work:
<img width="1552" alt="Screenshot 2020-12-29 at 9 45 43 PM" src="https://user-images.githubusercontent.com/31685197/103288174-4828b100-4a1f-11eb-8cff-aa9f21cbfb35.png">

**Reply by kiwicopple** [reply]
Thanks @soedirgo - I couldn't figure out what was happening on codesandbox before I went away. 

@jvbraganca is it working for you now?

**Reply by jvbraganca** [reply]
Hello again @kiwicopple, on Sandbox yes but tried on the real application and couldn’t make it work.
On Monday I will try to make a new database and see if it works.

**Reply by jvbraganca** [reply]
@kiwicopple, I made another DB, and it's working now. Thanks for the support! I wonder what the problem was.

**Comment by j2l** [comment]
May I join because I have the same problem?
I'm trying subscription with node-red and nothing is received despite a valid getSubscriptions() result.
My setup is simple:
![image](https://user-images.githubusercontent.com/65325/106142516-6e2c9880-6126-11eb-85cb-52570121e2a8.png)
When I manually inject an update, the subscribe code (see below) is not triggered.
```
const supabase = createClient('https://dbmbrmmqenkdnscjwslk.supabase.co', 'eyJhbG...ciOiJ4');
const tasks = supabase
  .from('*')
  .on('*', payload => {
    msg.payload = 'working', payload})
  .subscribe();
```
while ` msg.payload = supabase.getSubscriptions();` results in:
```
[{"topic":"realtime:public","params":{},"socket":{"channels":"[Circular ~.payload]","endPoint":"wss://dbmbrmmqenkdnscjwslk.supabase.co/realtime/v1/websocket","headers":{},"params":{"apikey":"ey4"},"timeout":10000,"heartbeatIntervalMs":30000,"longpollerTimeout":20000,"pendingHeartbeatRef":null,"ref":1,"conn":{"_listeners":{},"_url":"wss://dbmbrmmqenkdnscjwslk.supabase.co/realtime/v1/websocket?apikey=eyJh4&vsn=1.0.0","_readyState":0,"_extensions":"","_bufferedAmount":0,"_binaryType":"arraybuffer","_client":{"_events":{},"_eventsCount":2,"config":{"maxReceivedFrameSize":1048576,"maxReceivedMessageSize":8388608,"fragmentOutgoingMessages":true,"fragmentationThreshold":16384,"webSocketVersion":13,"assembleFragments":true,"disableNagleAlgorithm":true,"closeTimeout":5000,"tlsOptions":{}},"protocols":[],"origin":null,"url":{"protocol":"wss:","slashes":true,"auth":null,"host":"dbmbrmmqenkdnscjwslk.supabase.co","port":"443","hostname":"dbmbrmmqenkdnscjwslk.supabase.co","hash":null,"search":"?apikey=eyJh4&vsn=1.0.0","query":"apikey=eyJh84&vsn=1.0.0","pathname":"/realtime/v1/websocket","path":"/realtime/v1/websocket?apikey=eyJh4&vsn=1.0.0","href":"wss://dbmbrmmqenkdnscjwslk.supabase.co/realtime/v1/websocket?apikey=eyJhbGc4&vsn=1.0.0"},"secure":true,"base64nonce":"63L671H5err1JHcftXlhsQ=="}},"sendBuffer":[null],"stateChangeCallbacks":{"open":[],"close":[],"error":[],"message":[]},"reconnectTimer":{"tries":0}},"bindings":[{"event":"phx_close"},{"event":"phx_error"},{"event":"phx_reply"},{"event":"*"},{"event":"phx_error"},{"event":"phx_close"},{"event":"chan_reply_1"}],"state":"joining","joinedOnce":true,"pushBuffer":[],"timeout":10000,"joinPush":{"channel":"[Circular ~.payload.0]","event":"phx_join","payload":{},"timeout":10000,"sent":true,"timeoutTimer":{"_called":false,"_idleTimeout":10000,"_idlePrev":{"_idleNext":"[Circular ~.payload.0.joinPush.timeoutTimer]","_idlePrev":"[Circular ~.payload.0.joinPush.timeoutTimer]","_unrefed":false,"msecs":10000,"_timer":{"_list":"[Circular ~.payload.0.joinPush.timeoutTimer._idlePrev]"}},"_idleNext":{"_idleNext":"[Circular ~.payload.0.joinPush.timeoutTimer]","_idlePrev":"[Circular ~.payload.0.joinPush.timeoutTimer]","_unrefed":false,"msecs":10000,"_timer":{"_list":"[Circular ~.payload.0.joinPush.timeoutTimer._idleNext]"}},"_idleStart":28330826,"_repeat":null,"_destroyed":false},"ref":"1","receivedResp":null,"recHooks":[{"status":"ok"},{"status":"timeout"},{"status":"ok"},{"status":"error"},{"status":"timeout"}],"refEvent":"chan_reply_1"},"rejoinTimer":{"tries":0}}]
```
db was made using your todo example, I'm just trying to adapt it to node-red and maybe create a supabase module for node-red :)

`CREATE PUBLICATION supabase_realtime FOR ALL TABLES;`
throws an error about `supabase_realtime  already exists`

And `select * from pg_replication_slots;`:
![image](https://user-images.githubusercontent.com/65325/106144005-6ff75b80-6128-11eb-9c32-1ff190c77c26.png)

My server is behind a NAT, but websocket was working before.

I guess, if it happens from time to time, maybe there's something to investigate. I can leave the db as is if you'd like.
If you don't have time, let me know and I'll try with another db.

BTW, where are logs on the server? Because I had a hard time debugging malformed CRUD functions and nothing was sent back to my client, except when it was working of course. I noticed that id jumped from 3 to 6, so I know the server got some malformed message, but no way to debug them.

Thank you!

**Reply by kiwicopple** [reply]
Hey J2l, the code looks fine but so it's a bit hard to determine where the problem lies with this one - it could be NodeRed, the NAT, your database, or our infra. Is there a way you can isolate the problem a bit? Could you see if the realtime updates are getting received on your local machine?

**Comment by j2l** [comment]
@kiwicopple thank you for your reply,
What code is needed to check the realtime updates?

**Comment by laurentS** [comment]
@kiwicopple I have hit this problem again.
Here is what I did that led to the notifications not coming through (I think):
```sql
delete from pg_publication where pubname='supabase_realtime';
-- some data upload that would not complete while the publication was active
create publication supabase_realtime for all tables;
```

I don't expect any of the other SQL commands to have caused any issue [eg. create table, create view, insert...].

After this, notifications are not coming through anymore.
Here are some results I'm seeing in the DB:
```sql
postgres=# select * from pg_publication;
  oid  |      pubname      | pubowner | puballtables | pubinsert | pubupdate | pubdelete | pubtruncate
-------+-------------------+----------+--------------+-----------+-----------+-----------+-------------
 33052 | supabase_realtime |       10 | t            | t         | t         | t         | t
(1 row)
```
and
```sql
postgres=# select * from pg_replication_slots ;
 slot_name |  plugin  | slot_type | datoid | database | temporary | active | active_pid | xmin | catalog_xmin | restart_lsn | confirmed_flush_lsn
-----------+----------+-----------+--------+----------+-----------+--------+------------+------+--------------+-------------+---------------------
 realtime  | pgoutput | logical   |  13398 | postgres | f         | f      |            |      |         1337 | 0/2CE41538  | 0/2CE41538
(1 row)
```
I am puzzled by the fact that `active_pid` is null. All the examples above have some value there. Is there some service/daemon that did not restart/reconnect?

**Reply by laurentS** [reply]
Might be helpful to know that I'm seeing `wss://databaseid.supabase.co/realtime/v1/websocket?apikey=eyJhbGciOiJIUzI1NiIsInRxxxxxxxxxxxxxxxCnRQYo&vsn=1.0.0` returning `502
Bad Gateway`

**Reply by inian** [reply]
Hi @laurentS, yes I see the issue is because realtime was crashing because the publication `supabase_realtime` does not exist. The [recommended way](https://supabase.io/docs/guides/auth#disable-realtime-for-private-tables) to disable realtime temporarily is to create an empty publication which doesn't subscribe to any table. 

I have restarted the realtime server now. Can you check if this fixes your problem?

**Reply by laurentS** [reply]
Hi @inian yes, it's working now, thanks! I'll change my code to follow the recommendations you linked to. As a follow up question, would it make sense to have the realtime server try to restart every so often if it crashes? I don't really understand the implications, but my publication was gone for about a minute max and that rendered the server unusable for me. I understand I made a mistake in my code, but I'm willing to bet it's not going to be the last time (for me or others). It feels overkill to have to reach out to you guys for something like this. Alternatively, would you be able to have a "restart realtime server" button on the web UI?

**Comment by epomeranz** [comment]
Hi I'm having the same issue with real-time but in flutter (exact flutter version below).
My code, database, etc all looks the same except that is in flutter

When running the app the socket connects ok, and then I receive the following "onMessage" events from the socket connection:
1. Once, right after creating the subscription: 
  -{event: phx_reply, payload: {response: {}, status: ok}, ref: , topic: realtime:public}
2. Every few seconds: 
  -{event: phx_reply, payload: {response: {reason: unmatched topic}, status: error}, ref: 1, topic: phoenix}
  -{event: phx_reply, payload: {response: {reason: unmatched topic}, status: error}, ref: 2, topic: phoenix}
  -{event: phx_reply, payload: {response: {reason: unmatched topic}, status: error}, ref: 3, topic: phoenix}


Flutter 2.3.0-1.0.pre • channel dev • https://github.com/flutter/flutter.git
Framework • revision d97f41caed (2 weeks ago) • 2021-04-30 12:35:21 -0700   
Engine • revision e7939e091e
Tools • Dart 2.14.0 (build 2.14.0-48.0.dev)
 
How can I check if my real-time server needs restarting? 

Thanks!

**Reply by tonyborchert** [reply]
I made a completely new project and this also happened to me

**Comment by muhammad0071** [comment]
hey, i have an issue with hosted real-time local, got empty data when I run `select * from pg_replication_slots;` I use Postgres 14.7 btw, not from superbase, is that something I missed? `walt2json` extension not installed on my Postgres also
